synapse/handlers/relations.py

```diff
@@ -13,7 +13,16 @@
 # limitations under the License.
```

Comment:
Using the artificial Complement test from https://github.com/matrix-org/complement/pull/443 which has zero relations, `get_bundled_aggregations` has gone from ~1000ms -> ~600ms (500-700ms) with this PR 📉

All of the `get_annotations_for_event` calls have been boiled down to a single `get_annotations_for_events` 💪 and @clokep suspects he can do the same thing for the `m.reference` (which is the remaining `get_relations_for_event` bulk)

Trace JSON (drag and drop onto Jaeger under the `JSON File` tab): https://gist.github.com/MadLittleMods/b1fb31f5527b90f59d69111f8f2c2a8b


<img src="https://user-images.githubusercontent.com/558581/202822139-462a1e46-ed9a-4d04-9486-543aad9ac4cc.png" width="300">


To reproduce my results:

<details>
<summary>Expand for reproduction steps</summary>

Have Jaeger running locally to collect the spans/traces. If you don't have one, you can spin one up with Docker:
```
docker run -d --name jaeger \
  -e COLLECTOR_ZIPKIN_HOST_PORT=:9411 \
  -e COLLECTOR_OTLP_ENABLED=true \
  -p 6831:6831/udp \
  -p 6832:6832/udp \
  -p 5778:5778 \
  -p 5775:5775/udp \
  -p 16686:16686 \
  -p 4317:4317 \
  -p 4318:4318 \
  -p 14250:14250 \
  -p 14268:14268 \
  -p 14269:14269 \
  -p 9411:9411 \
  jaegertracing/all-in-one:1.35
```

Get the right Synapse code in place:

```
# Use the OpenTelemetry PR as a base so we can trace in the Complement Docker tests
git checkout madlittlemods/11850-migrate-to-opentelemetry
# Create a new scratch branch
git checkout -b madlittlemods/test-14491-faster-scratch
# Merge this PR in
git merge clokep/batch-aggs
```

Add the following config to `docker/conf/homeserver.yaml#L189-L206`

https://github.com/matrix-org/synapse/blob/04de9ea73809670f4ce2cf0dc085065a324453e8/docker/conf/homeserver.yaml#L189-L206

Go to your Complement checkout:
```
git checkout madlittlemods/room-messages

# This will take a while to run the first time while it sets up the massive amount of rooms, senders, and messages on `hs1`.
TEST_ONLY_IGNORE_POETRY_LOCKFILE=1 TEST_ONLY_SKIP_DEP_HASH_VERIFICATION=1 COMPLEMENT_DEBUG=1 COMPLEMENT_KEEP_BLUEPRINTS="fed.perf_many_messages.hs1" COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh -run TestMessagesOverFederation -p 1
```

Visit http://localhost:16686/search?limit=20&lookback=1h&operation=RoomMessageListRestServlet&service=hs2%20master to find the trace in Jaeger


</details>